### PR TITLE
fix(replay): New replay foundation for hydration errors

### LINKED ISSUE #75073: [Hydration Errors] [Bug] Bug with Diff Tooling Timestamps
More details here: https://sentry.sentry.io/feedback/?referrer=slack&notification_uuid=6045b4fe-9e53-40b6-9a94-b318c5fa674a&alert_rule_id=15210908&alert_type=issue&feedbackSlug=javascript%3A5641496707&project=11276

### LINKED ISSUE #75689: [Hydration Errors] [Bug] Before in HTML Diff is Empty When There is Content Before in Visual Diffs
Reported by 3 users:
- [JAVASCRIPT-2V5B](https://sentry.sentry.io/feedback/?referrer=github_integration&feedbackSlug=javascript%3A5682184590&project=11276) 
- [JAVASCRIPT-2V6P](https://sentry.sentry.io/feedback/?feedbackSlug=javascript%3A5691053354&project=11276)
- [JAVASCRIPT-2V8B](https://sentry.sentry.io/feedback/?feedbackSlug=javascript%3A5708356821&project=11276)

### LINKED ISSUE #76079: [Hydration Errors] [Bug] HTML Diff is the same (before and after) when it should not be
Details [here on feedback report](https://sentry.sentry.io/feedback/?alert_rule_id=15210908&alert_type=issue&feedbackSlug=javascript%3A5690972194&project=11276&query=assigned%3Ame&referrer=feedback_list_page&statsPeriod=14d) and [this one](https://sentry.sentry.io/feedback/?feedbackSlug=javascript%3A5691053354&project=11276)

### PR DESCRIPTION
**Pull request by ryan953**
Two problems I wanted to fix were:

1. Too many re-renders. With the old system all our replayer state was in one ginat context, so whenever anything changed (like the currentTime react would re-calculate everything on the page and see if it needed to be re-rendered. No longer! This new system splits state into multiple parts based on stability.
2. The previous <ReplayPlayer> component was doing too much. Specifically the measuring/scaling code in there was having trouble (the extra rendering from #1 above didn’t help). This was a root cause for why hydration errors rendered unreliably. Now we have a new, slimmer <ReplayPlayerContainment> component to do the measurements and scale the replay into the webpage. All the other responsibilities will be easily handled in some future PR.

This PR adds a bunch of react-hooks & providers to split up the state and make managing the `new Replayer` instances cleaner. This will be something to continue iterating on over time. For now using it to do hydration error stuff is a good stepping stone to get the code in and working in a scenario where the replay doesn't need to play, and doesn't need to deal with mobile replays. 

This rendering fix should solve these:
Fixes https://github.com/getsentry/sentry/issues/76079
Fixes https://github.com/getsentry/sentry/issues/75689
Fixes https://github.com/getsentry/sentry/issues/75073

### CODE REVIEW

File: static/app/components/replays/player/replayPlayerContainment.stories.tsx [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+    function Example() {
+      return (
+        <Fragment>
+          <p>
+            The size of the replay can be constrained to the width of the container,
+            either the page or a div.
+          </p>
+          <p>
+            Portrait mode captures will very likely be a problem, because devices like
+            iPhone have high pixel density the capture can be 800px tall, or more. And
```
Comment (michellewzhang) [inline static/app/components/replays/player/replayPlayerContainment.stories.tsx:57]:
```suggestion
            iPhone have high pixel density. For these devices, the capture can be 800px tall or more.
```

File: static/app/components/replays/player/replayPlayerContainment.stories.tsx [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+            zoomed to 1.5 magnification, resulting in 1,200 pixel tall replays!
```
Comment (michellewzhang) [inline static/app/components/replays/player/replayPlayerContainment.stories.tsx:58]:
```suggestion
            Zoomed to 1.5 magnification, this can result in 1,200 pixel tall replays!
```

File: static/app/components/replays/player/replayPlayerContainment.tsx [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+   *    the replay.
+   *    ie: with a wrapper that has CSS `height` set.
+   *
+   * "width"
+   *    Height available will not be measured. The replay will be
+   *   resized to accomodate the width available, and the height grows to
+   *   maintain the aspect-ratio. This can result in really tall replays when
+   *   captured on a mobile device.
+   *
+   * There is no "height" option, document flow doesn't usually expand in an
```
Comment (michellewzhang) [inline static/app/components/replays/player/replayPlayerContainment.tsx:30]:
```suggestion
   * There is no "height" option; document flow doesn't usually expand in an
```

File: static/app/components/replays/player/replayPlayerContainment.tsx [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+  children: (styles: CSSProperties) => ReactNode;
+
+  /**
+   * How to measure and resize the player.
+   *
+   * "both"
+   *    The default. Measure the player fully. Resize the player to fit within
+   *    the given width & height. Useful if you want to reserve some space for
+   *    the replay.
+   *    ie: with a wrapper that has CSS `height` set.
```
Comment (michellewzhang) [inline static/app/components/replays/player/replayPlayerContainment.tsx:22]: should we add a note about which one (`both` vs `width`) is used most often (similar to the note you made in the story)?

File: static/app/utils/replays/playback/providers/useReplayPlayerState.tsx [hunk 5]
```diff
@@ -0,0 +1,295 @@
...
+      type: 'setConfigIsSkippingInactive';
+    }
+  | {playbackSpeed: ReplayPrefs['playbackSpeed']; type: 'setConfigPlaybackSpeed'}
+  | {offsetMs: number; type: 'jumpToOffset'};
+
+interface State {
+  currentSpeed: undefined | number;
+  dimensions: Dimensions;
+  isFinished: boolean;
+  playerState: 'playing' | 'paused' | 'live';
```
Comment (michellewzhang) [inline static/app/utils/replays/playback/providers/useReplayPlayerState.tsx:43]: what's the `live` state?

File: static/app/utils/replays/playback/providers/useReplayPlayerState.tsx [hunk 6]
```diff
@@ -0,0 +1,295 @@
...
+      };
+    case 'didSkipStart':
+      return {...state, speedState: 'skipping', currentSpeed: replayerAction.speed};
+    case 'didSkipEnd':
+      return {...state, speedState: 'normal', currentSpeed: replayerAction.speed};
+    case 'didFlush':
+      return state;
+
+    case 'didPlayerStateChange':
+      // console.log('didPlayerStateChange', state);
```
Comment (michellewzhang) [inline static/app/utils/replays/playback/providers/useReplayPlayerState.tsx:159]: extra comment

File: static/app/components/replays/player/replayPlayer.tsx [hunk 7]
```diff
@@ -0,0 +1,102 @@
...
+      });
+    }
+  }, [prefs, userAction]);
+
+  return mountPointRef;
+}
+
+interface Props extends HTMLAttributes<HTMLDivElement> {
+  css?: Interpolation<Theme>;
+  offsetMs?: undefined | number;
```
Comment (billyvg) [inline static/app/components/replays/player/replayPlayer.tsx:82]:
```suggestion
  offsetMs?: number;
```

File: static/app/components/replays/player/replayPlayerContainment.tsx [hunk 8]
```diff
@@ -0,0 +1,84 @@
...
+   *   maintain the aspect-ratio. This can result in really tall replays when
+   *   captured on a mobile device.
+   *
+   * There is no "height" option, document flow doesn't usually expand in an
+   * uncontrolled way like height does.
+   */
+  measure?: 'both' | 'width';
+}
+
+export default function ReplayPlayerContainment({children, measure = 'both'}: Props) {
```
Comment (billyvg) [inline static/app/components/replays/player/replayPlayerContainment.tsx:36]: Maybe a name related to what it's doing irt dimensions

```suggestion
export default function ReplayPlayerSizer({children, measure = 'both'}: Props) {
```

```suggestion
export default function ReplayPlayerMeasurer({children, measure = 'both'}: Props) {
```

```suggestion
export default function ReplayPlayerDimension({children, measure = 'both'}: Props) {
```

File: static/app/components/replays/player/styles.tsx [hunk 9]
```diff
@@ -18,7 +18,9 @@ export const baseReplayerCss = css`
   .replayer-wrapper > iframe {
     border: none;
     background: white;
+  }
 
+  &[data-inspectable='true'] .replayer-wrapper > iframe {
```
Comment (billyvg) [inline static/app/components/replays/player/styles.tsx:23]: intentional?

File: static/app/utils/replays/playback/hooks/useReplayCurrentTime.tsx [hunk 10]
```diff
@@ -0,0 +1,37 @@
...
+    {enabled: Boolean(replayer) && playerState === 'playing'}
+  );
+
+  useEffect(() => {
+    if (state?.value === 'paused' && state?.context.timeOffset !== undefined) {
+      if (isFinished && replayer) {
+        callback({timeMs: replayer?.getCurrentTime()});
+      } else {
+        callback({timeMs: state?.context.timeOffset});
+      }
```
Comment (billyvg) [inline static/app/utils/replays/playback/hooks/useReplayCurrentTime.tsx:30]: lgoaf low, but this helps to see what is conditional

```suggestion
      callback({
        timeMs: isFinished && replayer
          ? replayer?.getCurrentTime()
          : state?.context.timeOffset
      });
```

File: static/app/utils/replays/playback/hooks/useReplayPlayWhenVisible.tsx [hunk 11]
```diff
@@ -0,0 +1,24 @@
+import {useEffect} from 'react';
+
+import useReplayPlayerState, {
+  useReplayUserAction,
+} from 'sentry/utils/replays/playback/providers/useReplayPlayerState';
+
+export default function useReplayPlayWhenVisible() {
```
Comment (billyvg) [inline static/app/utils/replays/playback/hooks/useReplayPlayWhenVisible.tsx:7]: Doesn't this pause when not visible?

**Review by billyvg — COMMENTED** [review]
We should probably split off some of the changes that aren't related to the player (e.g. duration, openReplayComparisonButton)

File: static/app/components/duration/duration.tsx
Comment (ryan953) [inline static/app/components/duration/duration.tsx:1]: This file will land in https://github.com/getsentry/sentry/pull/76711 first.

Comment (ryan953) [inline static/app/components/duration/duration.tsx:1]: This file will be merged as part of https://github.com/getsentry/sentry/pull/76711 first 🤞

File: static/app/components/replays/player/__stories__/jumpToOffsetButtonBar.tsx [hunk 12]
```diff
@@ -0,0 +1,34 @@
...
+        return (
+          <Button
+            key={interval}
+            onClick={() => userAction({type: 'jumpToOffset', offsetMs: intervalMs})}
+            size="sm"
+          >
+            {formatDuration({
+              duration: [intervalMs, 'ms'],
+              style: 'h:mm:ss',
+              precision: 'sec',
```
[the remark this answers is not in the document]
Comment (ryan953) [reply static/app/components/replays/player/__stories__/jumpToOffsetButtonBar.tsx:27]: for stories i think it's fine. The component is for adding some static buttons that work kind of how the timeline works. I don't think the extra precision is important for stories.

SCR-20240904-oivz

File: static/app/components/replays/player/__stories__/replaySlugChooser.tsx [hunk 13]
```diff
@@ -0,0 +1,42 @@
...
+function LoadReplay({children, replaySlug}: {children: ReactNode; replaySlug: string}) {
+  const organization = useOrganization();
+  const {replay, fetching} = useReplayReader({
+    orgSlug: organization.slug,
+    replaySlug,
+  });
+
+  if (!replay || fetching) {
+    return 'Loading...';
+  }
```
[the remark this answers is not in the document]
Comment (ryan953) [reply static/app/components/replays/player/__stories__/replaySlugChooser.tsx:39]: can do! but nothing fancy for the stories environment.

File: static/app/components/replays/player/replayPlayer.stories.tsx [hunk 14]
```diff
@@ -0,0 +1,50 @@
...
+
+export default storyBook(ReplayPlayer, story => {
+  story('Default', () => {
+    function Example() {
+      return (
+        <SideBySide>
+          <ReplayPlayPauseButton />
+          <ReplayCurrentTime />
+          <ReplayPreferenceDropdown speedOptions={[0.5, 1, 2, 8]} />
+          <JumpToOffsetButtonBar intervals={['0m', '1ms', '1m', '8m', '12m']} />
```
[the remark this answers is not in the document]
Comment (ryan953) [reply static/app/components/replays/player/replayPlayer.stories.tsx:23]: oh damn, maybe 1ms is important to show.

File: static/app/components/replays/player/__stories__/jumpToOffsetButtonBar.tsx (hunk 12 above)
[the remark this answers is not in the document]
Comment (ryan953) [reply static/app/components/replays/player/__stories__/jumpToOffsetButtonBar.tsx:27]: oh, saw this: https://github.com/getsentry/sentry/pull/76927/files#r1744407931

File: static/app/components/replays/player/replayPlayer.tsx (hunk 7 above)
Comment (ryan953) [reply static/app/components/replays/player/replayPlayer.tsx:82]: that's not the same!

`<ReplayPlayer offsetMs={undefined} />` is allowed when the type is `offsetMs?: undefined | number;` 

but more likely there will be a variable passed along to set the value: `<ReplayPlayer offsetMs={offsetMs} />`

File: static/app/components/replays/player/replayPlayerContainment.tsx (hunk 4 above)
Comment (ryan953) [reply static/app/components/replays/player/replayPlayerContainment.tsx:22]: hrm, i probably shouldn't assume what's more common, but better explain the context for picking each.

the `width` option is good for the replay page itself because the height is flexible.

File: static/app/components/replays/player/styles.tsx (hunk 9 above)
Comment (ryan953) [reply static/app/components/replays/player/styles.tsx:23]: yeah. i should add a story for this

I remember we allowed pointer events, but in the app today they're disabled for some reason, the styles might be intentionally or unintentionally messed up, i can't sort out which... but pointer-event is effectively `none` in all the places afaik.

I split this out on purpose in case we wanted to allow pointer-events again.

File: static/app/utils/duration/formatDuration.tsx
Comment (ryan953) [inline static/app/utils/duration/formatDuration.tsx:1]: will be part of https://github.com/getsentry/sentry/pull/76711

File: static/app/utils/replays/playback/providers/useReplayPlayerState.tsx (hunk 5 above)
Comment (ryan953) [reply static/app/utils/replays/playback/providers/useReplayPlayerState.tsx:43]: it's when events are streaming in instead of being all pre-loaded up front.

we don't use this mode, i just ported over all the types from the rrweb package for completeness.

File: static/app/utils/replays/types.tsx [hunk 15]
```diff
@@ -17,6 +17,11 @@ import invariant from 'invariant';
 
 import type {HydratedA11yFrame} from 'sentry/utils/replays/hydrateA11yFrame';
 
+export type Dimensions = {
```
[the remark this answers is not in the document]
Comment (ryan953) [reply static/app/utils/replays/types.tsx:20]: oops, yeah. This line has been following me around ugh

**Comment by billyvg** [comment]
I don't think we should add tests for the sake of coverage but the 0 tests is kind of 😬 are there are any spots where tests would make sense?

**Comment by michellewzhang** [comment]
> I don't think we should add tests for the sake of coverage but the 0 tests is kind of 😬 are there are any spots where tests would make sense?

agree, a test of sorts would be good

File: static/app/components/replays/diff/replayDiffChooser.tsx [hunk 16]
```diff
@@ -80,26 +77,6 @@ export default function ReplayDiffChooser({
           </TabPanels.Item>
         </TabPanels>
       </Flex>
-    </TabStateProvider>
-  );
-}
-
-// TODO(ryan953): put this into static/app/components/tabs/index.tsx near `<Tabs>`
-function TabStateProvider<T extends string | number>({
```
Comment (michellewzhang) [inline static/app/components/replays/diff/replayDiffChooser.tsx]: why was this added then deleted🤔

File: static/app/utils/replays/playback/hooks/useReplayPlayWhenVisible.tsx [hunk 17]
```diff
@@ -1,24 +0,0 @@
-import {useEffect} from 'react';
```
Comment (michellewzhang) [inline static/app/utils/replays/playback/hooks/useReplayPlayWhenVisible.tsx]: same here -- what was this file used for? looks like it got deleted

**Review by michellewzhang — APPROVED** [review]
some comments about changes but otherwise lgtm, would be great to add a test too!

File: static/app/utils/replays/playback/hooks/useReplayPlayWhenVisible.tsx (hunk 11 above)
Comment (ryan953) [reply static/app/utils/replays/playback/hooks/useReplayPlayWhenVisible.tsx]: I don't think that rrweb pauses on it's own, we have code for that in replayContext.tsx

File: static/app/components/replays/diff/replayDiffChooser.tsx (hunk 16 above)
Comment (ryan953) [reply static/app/components/replays/diff/replayDiffChooser.tsx]: it's more UI focused, i will followup and make the Modal better, but not in this PR

File: static/app/utils/replays/playback/hooks/useReplayPlayWhenVisible.tsx (hunk 17 above)
Comment (ryan953) [reply static/app/utils/replays/playback/hooks/useReplayPlayWhenVisible.tsx]: It's not used yet. There's no play/pause for hydration errors so it's kind of an extra unused file at this point. It's related to the larger replayContext.tsx refactor idea, so it'll come back later.

File: static/app/utils/replays/playback/hooks/useReplayPlayWhenVisible.tsx (hunk 11 above)
Comment (billyvg) [reply static/app/utils/replays/playback/hooks/useReplayPlayWhenVisible.tsx]: I mean this hook, look at line 14

Comment (ryan953) [reply static/app/utils/replays/playback/hooks/useReplayPlayWhenVisible.tsx]: oh, it's named backwards! derp took me long enough

Comment (ryan953) [reply static/app/utils/replays/playback/hooks/useReplayPlayWhenVisible.tsx]: also it's not part of this diff anymore, because nothing plays for hydration errors. i'll note it over here which is sort of the OG branch: https://github.com/getsentry/sentry/pull/75175/files#r1747319889
